### ISSUE TITLE
False positive

### ISSUE THREAD

**Issue by wutianba** [NASA-SW-VnV/ikos#94]
Using DBM abstraction domain, such data read from memory cannot be related to local variables.
```
int main() {
  int gParaGyroUsed[6]; 
  int buf32[100];
  if (buf32[2] < 6) {
    gParaGyroUsed[buf32[2]] = 1;
  }
  return 0;
}
```

**Comment by arthaud** [comment]
Hi @wutianba,

This is not a false positive. `buf32[2]` could be negative and thus produce a buffer overflow.

Using the following code, all checks pass:
```c
int main() {
  int gParaGyroUsed[6]; 
  int buf32[100];
  if (buf32[2] >= 0 && buf32[2] < 6) {
    gParaGyroUsed[buf32[2]] = 1;
  }
  return 0;
}
```

**Comment by wutianba** [comment]
This is a typo. The code below will have a false positive.
```
int main() {
  unsigned int gParaGyroUsed[6]; 
  unsigned int buf32[100];
  if (buf32[2] < 6) {
    gParaGyroUsed[buf32[2]] = 1;
  }
  return 0;
}
```

**Comment by arthaud** [comment]
Confirmed.

This is an issue that I am well aware of. The problem is that our memory abstraction uses symbolic cell variables such as `C{location,offset,size}`. These variables need to have a fixed type for the abstract domains. I decided to type them as signed integers since this is more common. Because of this, we need to perform conversions when reading/writing in memory.

For instance, the first code gives this set of constraints:
```
x = (unsigned int)C{buf32,8,4}
assert x < 6
y = (unsigned int)C{buf32,8,4}
check y < 6
```

Since the current analysis only does forward propagation, the DBM domain cannot infer a relation between `x` and `C{buf32,8,4}` when executing the first statement. Remember, the DBM domain only infers properties of the form `x - y <= C`.

To fix this, we could either do backward propagation, or change the memory abstract domain to have dynamically typed cells. I don't know how hard it would be to implement this.

**Comment by wutianba** [comment]
Since LLVM IR is in the form of SSA, the same variable does not appear in multiple operations.
Can we think that if `a` is signed, the relationship of `b = (unsigned)a` is
```
if (a.l > 0)
   (b = a) -> (a - b >= 0, b - a >= 0)
else
   (b = (UNSINGED_MAX + a) or a) -> (b - a >= UNSINGED_MAX, a - b >= 0)
```

**Comment by arthaud** [comment]
This is too complicated for any numerical abstract domain.

**Comment by arthaud** [comment]
If you have ideas to fix it, I would love to hear it.
I didn't understand what "tricks" you are referring to.

**Comment by wutianba** [comment]
`core\include\ikos\domains\dbm.hpp:1014` in `IKOS 1.3`
```c++
  void apply(conv_operation_t op,
             VariableName x,
             VariableName y,
             uint64_t from,
             uint64_t to) {
    // Requires normalization
    normalize();

    if (is_bottom())
      return;

    interval_t v_x = interval_t::top();
    interval_t v_y = to_interval(y);

    switch (op) {
      case OP_TRUNC: {
        v_x = v_y.Trunc(from, to);
        break;
      }
      case OP_ZEXT: {
        v_x = v_y.ZExt(from, to);
        break;
      }
      case OP_SEXT: {
        v_x = v_y.SExt(from, to);
        break;
      }
      default: { ikos_unreachable("invalid operation"); }
    }
```
Get `a.l` or `a.u` by DBM[a, 0] or DBM[0, a] and add relation below for OP_ZEXT may solve it.
```
if (a.l > 0)
   (b = a) -> (a - b >= 0, b - a >= 0)
else
   (b = ((UNSINGED_MAX + a) or a) -> (b - a >= UNSINGED_MAX, a - b >= 0)
```

**Comment by arthaud** [comment]
Good idea @wutianba.

I applied this in 22213e847d8614d6a168c4c144f21f96e21886a1 but it doesn't solve the false positive.

IKOS is now able to infer that `buf[2] < 6`.
```
x = (unsigned int)buf[2]
assert x < 6
```
is translated into:
```
-2**31 <= buf[2] <= 2**31-1
x = mod(buf[2], 2**32)
0 <= x <= 2**32-1
assert x < 6
```
ikos infers:
```
x = mod(buf[2], 2**32)
-> buf[2] <= x <= buf[2] + 2**32
assert x < 6
-> buf[2] <= x < 6
```

Unfortunately, it still cannot infer that `buf[2] >= 0`.

**Comment by arthaud** [comment]
This is now fixed in 224fcee4b4c369666c7f50d84097247294d70b64.

You will have to use a relational abstract domain.
For instance, you can use `dbm` or `apron-ppl-polyhedra`.
For bigger code, I would recommend `var-pack-dbm`. It scales pretty well and has a good precision.
